### PR TITLE
Update active learning loop

### LINKED ISSUE #9: Finish Active Learning script + documentation/readme
Effort: 2 people (max 3)

First attempt of an active learning loop with connection to Rubrix for labelling was made in branch [active-learning-rubrix](https://github.com/EdAbati/fsdl-2022-weak-supervision-project/tree/active-learning-rubrix). See [this notebook](https://github.com/EdAbati/fsdl-2022-weak-supervision-project/blob/active-learning-rubrix/notebooks/rubrix-active-learning-small-text.ipynb)

Some parts are missing:
- use unlabelled data resulted from weak labelling
- decide on loop initialisation (probably using prediction of model created in #7)
- decide sampling strategy
- decide whether `small-text` should be used or we should code something from scratch
- decide if we can switch away from the `@listener` API to something more straightforward like [this example](https://rubrix.readthedocs.io/en/stable/tutorials/05-active_learning.html)

Also we should create a documentation explaining the design decisions, describing how an active learning loop could be used and deployed, etc.

FYI @bergr7

### PR DESCRIPTION
**Pull request by bergr7**
Added active learning loop with Rubrix and small-text. We decided to go for small-text in the end as creating everything from scratch was too much given the time constraints. Using small-text is easier but there are some limitations:

- Less control over the model: I couldn't find a way of plugging in the baseline model. I needed to initialize the model on the weakly supervised dataset instead. We could use the same hps as the ones we used for the baseline.
- Metrics monitoring: Not an easy way to integrate the test set. A bit of manual work is needed to integrate the active learning loop with the deployment strategy.
- Rubrix's listener is a bit confusing and you don't know what's going on at times.

Next (if time):
- More sophisticated query strategy
- Refactor notebook to prod ready code
- Integrate active learning loop with deployment strategy

Closes #9

### CODE REVIEW

File: services/jupyter/src/notebooks/rubrix/rubrix-active-learning-loop.ipynb [hunk 1]
```diff
@@ -0,0 +1,701 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 123,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# define query strategy\n",
+    "query_strategy = BreakingTies()"
```
Comment (alekshiidenhovi) [inline services/jupyter/src/notebooks/rubrix/rubrix-active-learning-loop.ipynb:520]: We could try maybe Entropy Based methods since they take into consideration probs for all classes, instead of just the two most probable in breaking ties

File: services/jupyter/src/notebooks/rubrix/rubrix-active-learning-loop.ipynb [hunk 2]
```diff
@@ -0,0 +1,701 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 130,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# this takes quite a bit of time... (1hr!!) - we initilize the active learner with the weak labels\n",
+    "indices_labeled = initialize_active_learner(active_learner, train_labels, weak_train_labels)"
```
Comment (alekshiidenhovi) [inline services/jupyter/src/notebooks/rubrix/rubrix-active-learning-loop.ipynb:558]: If I understood the documentation correctly, this line fine-tunes the model with the weak labels, right?

File: services/jupyter/src/notebooks/rubrix/rubrix-active-learning-loop.ipynb [hunk 3]
```diff
@@ -0,0 +1,701 @@
...
+    "ACCURACIES = []"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 134,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "@listener(dataset=DATASET_NAME,\n",
```
Comment (alekshiidenhovi) [inline services/jupyter/src/notebooks/rubrix/rubrix-active-learning-loop.ipynb:679]: I'm a bit confused how the @listener interacts with active_learning_loop and would prefer more glassbox solution, but I guess this is fine due to time constraints.

**Review by alekshiidenhovi — COMMENTED** [review]
Seems good to me, since you have been able to made it run on your machine. I would maybe opt out for entropy based querying strategy instead of margin sampling but we can discuss this later as well.

File: services/jupyter/src/pyproject.toml [hunk 4]
```diff
@@ -26,6 +26,12 @@ transformers = "^4.22.2"
 typer = "^0.6.1"
 wandb = "^0.13.3"
 boto3 = {version = "^1.24.89", source = "pypi"}
+rubrix = {extras = ["listeners"], version = "^0.18.0"}
+loguru = "^0.6.0"
+snorkel = {version = "^0.9.9", source = "pypi"}
+modAL = "^0.4.1"
+skorch = "^0.12.0"
```
Comment (EdAbati) [inline services/jupyter/src/pyproject.toml:33]: are these needed? If we just use `small-text` they can go I think

File: services/jupyter/src/notebooks/rubrix/rubrix-active-learning-loop.ipynb [hunk 5]
```diff
@@ -0,0 +1,701 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 64,
+   "metadata": {},
+   "outputs": [
+    {
+     "ename": "IndexError",
+     "evalue": "list index out of range",
+     "output_type": "error",
```
Comment (EdAbati) [inline services/jupyter/src/notebooks/rubrix/rubrix-active-learning-loop.ipynb]: There is an error in the output 👀

File: docker-compose.yml [hunk 6]
```diff
@@ -125,20 +126,24 @@ services:
...
     restart: unless-stopped
     ports:
       - "6900:80"
     environment:
       ELASTICSEARCH: http://elasticsearch:9200
       # RUBRIX_LOCAL_AUTH_USERS_DB_FILE: /config/.users.yaml
-      RUBRIX_LOCAL_AUTH_DEFAULT_PASSWORD: ${RUBRIX_LOCAL_AUTH_DEFAULT_PASSWORD-$$2y$$05$$X1i7R1R7CeFyD/dxZ5ttu.FcJQbL2Gh30.86OxCxzTXl/BR.RO.6O}
+      # RUBRIX_LOCAL_AUTH_DEFAULT_PASSWORD: ${RUBRIX_LOCAL_AUTH_DEFAULT_PASSWORD-$$2y$$05$$X1i7R1R7CeFyD/dxZ5ttu.FcJQbL2Gh30.86OxCxzTXl/BR.RO.6O}
     # volumes:
     #   - ./services/rubrix/src/config/.users.yaml:/config/.users.yaml
```
Comment (EdAbati) [inline docker-compose.yml:69]: can these be removed if unused?

File: docker-compose.yml [hunk 7]
```diff
@@ -125,20 +126,24 @@ services:
...
     networks:
       - rubrix
 
   elasticsearch:
+    env_file: .env
+    container_name: ${USER_NAME-test}-elasticsearch
```
Comment (EdAbati) [inline docker-compose.yml]:
```suggestion
```
Same can go

File: docker-compose.yml [hunk 8]
```diff
@@ -125,20 +126,24 @@ services:
     volumes:
       - streamlit-cachedata:/root/.cache
   rubrix:
+    env_file: .env
+    container_name: ${USER_NAME-test}-rubrix
```
Comment (EdAbati) [inline docker-compose.yml]:
```suggestion
```

You can remove this, I was playing around a bit 😅

File: services/jupyter/src/app/model/train_model.py [hunk 9]
```diff
@@ -17,7 +17,10 @@
 NUM_LABELS = 4
 
 
-def load_data(dataset_name: str = "bergr7/weakly_supervised_ag_news") -> tuple:
+def load_data(
+    dataset_name: str = "bergr7/weakly_supervised_ag_news",
+    split: bool = True,
+) -> tuple:
```
Comment (EdAbati) [inline services/jupyter/src/app/model/train_model.py]: This function could have been used in the 'active-learning' notebook.

**Review by EdAbati — CHANGES_REQUESTED** [review]
Just few suggestion below to clean up the branch a bit 🧹

Also 
- the content of the folder `notebook` is repeated inside `services/jupyter/src/notebooks/rubrix/` (judging by the history @diegoquintanav might have moved it). Which one is the most updated one? My guess that the content in `notebook` can go. 

- which notebook in `services/jupyter/src/notebooks/rubrix/` is the definitive one? my guess is that the ones ending with `-modal` and `-small-text` can go right?

File: services/jupyter/src/notebooks/rubrix/rubrix-active-learning-loop.ipynb (hunk 1 above)
Comment (bergr7) [reply services/jupyter/src/notebooks/rubrix/rubrix-active-learning-loop.ipynb:520]: Sure! I just went for the simplest one.

File: services/jupyter/src/notebooks/rubrix/rubrix-active-learning-loop.ipynb (hunk 2 above)
Comment (bergr7) [reply services/jupyter/src/notebooks/rubrix/rubrix-active-learning-loop.ipynb:558]: yeah exactly

File: services/jupyter/src/notebooks/rubrix/rubrix-active-learning-loop.ipynb (hunk 3 above)
Comment (bergr7) [reply services/jupyter/src/notebooks/rubrix/rubrix-active-learning-loop.ipynb:679]: Yeah  I agree but given the time constraints... I don't see any easy alternative I'm afraid

File: services/jupyter/src/pyproject.toml (hunk 4 above)
Comment (bergr7) [reply services/jupyter/src/pyproject.toml:33]: Diego created this branch and I work on top of it so I could use the setup. I guess we don't need that.

Comment (diegoquintanav) [reply services/jupyter/src/pyproject.toml:33]: yeah, I installed them to make some notebooks work. If we are not going to use them it's safe to remove them.

File: services/jupyter/src/notebooks/rubrix/rubrix-active-learning-loop.ipynb (hunk 5 above)
Comment (bergr7) [reply services/jupyter/src/notebooks/rubrix/rubrix-active-learning-loop.ipynb]: Yeah my bad. I created subsets for testing given that initializing the model takes quite a bit of time and I forgot to revert it back:

```
weak_train_text = [row["text"] for row in labeled_dataset["train"]][:100]
unlabeled_train_text = [row["text"] for row in unlabeled_dataset["unlabeled"]][:200]
# labels
weak_train_labels = [row["label"] for row in labeled_dataset["train"]][:100]
unlabeled_no_labels = [LABEL_UNLABELED for _ in unlabeled_train_text][:200]
```
I'll fix this.

File: docker-compose.yml (hunk 6 above)
Comment (diegoquintanav) [reply docker-compose.yml:69]: I set `team44` as the password, the default is `rubrix`. I'm happy with either.

Comment (bergr7) [reply docker-compose.yml:69]: @diegoquintanav could you answer this one?

Comment (diegoquintanav) [reply docker-compose.yml:69]: that is an environment variable that, when passed to the rubrix container (documented by rubrix in https://rubrix.readthedocs.io/en/stable/getting_started/user-management.html?highlight=password#how-to-override-the-default-user-password)

if not set then I think the default password is `1234`. We can remove it or keep it, I'm fine with either.

**Comment by bergr7** [comment]
@EdAbati Fixed the error in the notebook.

**Comment by EdAbati** [comment]
Resolved conflicts as discussed @bergr7 

@diegoquintanav can you have a look at this if you have time? :)

**Review by EdAbati — APPROVED** [review]

**Review by diegoquintanav — APPROVED** [review]
:+1:
